activerecord/lib/active_record/relation/query_methods.rb

```diff
@@ -1956,6 +1956,23 @@ def arel_columns(columns)
         end
       end
 
+      def arel_columns_from_array(columns, table_name = model.table_name)
+        columns.map do |column|
+          arel_column("#{table_name}.#{column}")
+        end
+      end
+
+      def arel_columns_from_hash(fields)
+        fields.flat_map do |table_name, columns|
+          case columns
+          when Array
+            arel_columns_from_array(columns, table_name)
+          when Symbol, String
+            arel_column(columns)
```

Comment:
We probably ought to raise on `else`. otherwise we'd silently ignore some types.

Reply:
Done.

Reply:
Thank you!